bundles/org.openhab.binding.unifiedremote/src/main/java/org/openhab/binding/unifiedremote/internal/UnifiedRemoteDiscoveryService.java

```diff
@@ -0,0 +1,187 @@
+            String name;
+            int tcpPort;
+            int udpPort;
+            String host;
+            String macAddress;
+            String publicIp;
+
+            ServerInfo(String host, int tcpPort, int udpPort, String name, String macAddress, String publicIp) {
+                this.name = name;
+                this.tcpPort = tcpPort;
+                this.udpPort = udpPort;
+                this.host = host;
+                this.macAddress = macAddress;
+                this.publicIp = publicIp;
+            }
+        }
+
+        /**
+         * Create a UDP socket on the service discovery broadcast port.
+         *
+         * @return open DatagramSocket if successful
+         * @throws RuntimeException if cannot create the socket
+         */
+        public DatagramSocket createSocket() throws SocketException {
+            DatagramSocket socket;
+            socket = new DatagramSocket();
+            socket.setBroadcast(true);
+            socket.setSoTimeout(TIMEOUT);
+            return socket;
+        }
+
+        private ServerInfo tryParseServerDiscovery(DatagramPacket receivePacket) throws ParseException {
+            String host = receivePacket.getAddress().getHostAddress();
+            String reply = new String(receivePacket.getData()).replaceAll("[\\p{C}]", NON_PRINTABLE_CHARTS_REPLACEMENT)
+                    .replaceAll("[^\\x00-\\x7F]", NON_PRINTABLE_CHARTS_REPLACEMENT);
+            if (!reply.startsWith(DISCOVERY_RESPONSE_PREFIX))
+                throw new ParseException("Bad discovery response prefix", 0);
+            String[] parts = Arrays
+                    .stream(reply.replace(DISCOVERY_RESPONSE_PREFIX, "").split(NON_PRINTABLE_CHARTS_REPLACEMENT))
+                    .filter((String e) -> e.length() != 0).toArray(String[]::new);
+            String name = parts[0];
+            int tcpPort = Integer.parseInt(parts[1]);
+            int udpPort = Integer.parseInt(parts[3]);
+            String macAddress = parts[2];
+            String publicIp = parts[4];
+            return new ServerInfo(host, tcpPort, udpPort, name, macAddress, publicIp);
+        }
+
+        /**
+         * Send broadcast packets with service request string until a response
+         * is received. Return the response as String (even though it should
+         * contain an internet address).
+         *
+         * @return String received from server. Should be server IP address.
+         *         Returns empty string if failed to get valid reply.
+         */
+        public void call(Consumer<ServerInfo> listener) {
+            byte[] receiveBuffer = new byte[MAX_PACKET_SIZE];
+            DatagramPacket receivePacket = new DatagramPacket(receiveBuffer, receiveBuffer.length);
+
+            DatagramSocket socket = null;
+            try {
+                socket = createSocket();
+            } catch (SocketException e) {
+                logger.error("Error creating socket: {}", e.getMessage());
```

Comment:
Bindings should only log to error if something severe happened, like the detection of a bug in your code. This could be debug. See this link for a description of the log levels: https://www.openhab.org/docs/developer/guidelines.html#f-logging